### ISSUE TITLE
[Bug] [zTrial] Bug has been reported in one of the samples

### ISSUE THREAD

**Issue by ddimatos** [ansible-collections/ibm_zos_core#714]
### Is there an existing issue for this?

- [X] There are no existing issues.

### Are the dependencies a supported version?

- [X] The dependencies are supported.

### IBM Z Open Automation Utilities

v1.1.1

### IBM Enterprise Python

v3.9.x

### IBM z/OS Ansible core Version

v1.4.0

### ansible-version

v2.11.x

### z/OS version

v2.5

### Bug description

A zTrial bug was reported, this is a place holder for more details to be provided.

**Comment by richp405** [comment]
This is the original issue email, from IBM UK team:

Walk through the directories to find copy_sort_fetch.yml and run it.

 

In this instance of Z Trial, the ansible.cfg file and the subfolders inside inventories including the inventory file itself has been pre-populated for your convenience. It is ready for you to run right out of the box!

All you have to do to run a playbook is navigate to the playbook directory and run the command ‘ansible-playbook -i inventories

Let’s try running the copy_sort_fetch playbook as an example. You can follow these same steps to run any other playbook provided in Z Trial. This playbook is contained in the data_transfer directory under the zos_concepts directory and under the main z_ansible_collections_samples directory

Find and read the README file that goes with the copy_sort_fetch playbook

To locate the copy_sort_fetch README file, in the explorer pane of VSCodium, click on the z_ansible_collections_sampledirectory to the view its contents, then click on the zos_concepts subdirectory, then click on the data_transfer subdirectory and lastly click on the copy_sort_fetch directory to expand and view the contents of the playbook directory which will include the README.md file. You can view the README in “Preview” mode inside VSCodium® by right-clicking on the README.md file and selecting ‘Open Preview’. Note, many of the links in the local Z Trial copy READMEs may be broken since connecting to external websites which is not permissible from within Z Trial.

You can view the README in “Preview” mode inside VSCodium® by right-clicking on the README.md file and selecting ‘Open Preview’. Note, many of the links in the local Z Trial copy READMEs may be broken since connecting to external websites which is not permissible from within Z Trial.

 

 

The ansible.cfg file is a source to configure Ansible® behavior. The file provided in the repo contains some default values suggested by the maintainers of the repo. For the local ansible.cg file, we have populated it with configurations needed to drive the playbook in the Z Trial environment.

The changelog.yml file contains a record of all notable changes to the playbook directory. This file serves little purpose to the playbook user but it is helpful in tracking changes to the repo over time.

Next up is the playbook itself.

The runtime-requirements.yml file lists out acceptable versions of Ansible for running the playbook. It also contains information about Ansible collection version requirements.

The final file, site.yml, is used in tracking whether or not requirements from the runtime-requirements.yml file is satisfied in the user system. It will not affect how a playbook is driven in Z Trial.

In general, for any playbook directory in the repo, the inventories directory needs to be modified and potentially the ansible.cfg file needs to be modified for a user to drive a sample playbook from the z_ansible_collections_samples repo.

In VSCodium, navigate to the playbook dir to run the playbook

Open up a terminal window in the same way as you have in prior scenarios (by clicking the ‘Terminal’ tab on the top ribbon and selecting the first option ‘New Terminal’.

Navigate to the playbook directory by calling

cd ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch

 

From here we are ready to run the playbook, just the same as in the other scenarios, run:

ansible-playbook -i inventories copy-sort-fetch.yml

 

Watch the playbook run!

 

The copy-sort-fetch playbook reads a file from inside the roles directory, transfers it to z/OS® for sorting, and then creates a file on the control node using zos_fetch, a control node refers to the machine which is driving the Ansible® playbook.

Feel free to modify and explore this playbook and the other playbooks included here to further understand the behavior of the tasks, ibm_zos_core modules, and the playbooks. Consider adding the Ansible pause task like we did in a previous scenario in different places to pause playbook execution and SSH into the z/OS machine to see the changes occuring on z/OS.

Le 29/03/2023 à 12:44, a écrit :

You have all the informations with detailed scenario in the notes attached.

Unfortunately i no longer have access to the ansible trial (expired) .

The playbook which fails is copy_sort_fetch in

z_ansible_collections_samples\zos_concepts\data_transfer\copy_sort_fetch

 

it is not in the trial instructions per say, but the instructions say that you can try some of the other available palybooks, and the failing playbook is copy_sort_fetch

see all my details below, it is self explanatory to anyone who understands this lab

The playbook has a sort step that fails.

in the playbook log you get an error message ' ICE218A 0 5 BYTE VARIABLE RECORD IS SHORTER THAN 65 BYTE MINIMUM FOR '

This is easily reproduceable, you just need to run this playbook, and you will get the error i show, with the solution i found and that i describe below

____________________________________________________________________________________________________________

One of the ansible playbooks fails : copy_sort_fetch

 

the SORT step fails with msg ICE218A 0 5 BYTE VARIABLE RECORD IS SHORTER THAN 65 BYTE MINIMUM FOR          FIELDS  (see the log further down)

The reason is that even though the temp dataset was allocated FB 80, the dataset was changed to VB 1028 ( which generates a SORT failure later), after this playbook step in main.yml :

 

- name: Copy local data file to remote z/OS temporary data set
  ibm.ibm_zos_core.zos_copy:
    src: "{{ role_path }}/files/SalaryData.txt"
    dest: "{{ dest_tmp_data_set }}"
  register: copy_result

 

Anyone trying this playbook will obviously get the same failure.

Has anyone reported it ?

How can it be fixed ?

 

 

dataset chars:

 Data Set Name . . . . : IBMUSER.P6843361.T0116742.C0000000                   
                                                                              
 General Data                           Current Allocation                    
  Management class . . : **None**        Allocated tracks  . : 1              
  Storage class  . . . : SCBASE          Allocated extents . : 1              
   Volume serial . . . : USER01                                               
   Device type . . . . : 3390                                                 
  Data class . . . . . : **None**                                             
   Organization  . . . : PS             Current Utilization                   
   Record format . . . : VB              Used tracks . . . . : 1              
   Record length . . . : 1028            Used extents  . . . : 1              
   Block size  . . . . : 6144                                                 
   1st extent tracks . : 1                                                    
   Secondary tracks  . : 1              Dates                                 
   Data set name type  :                 Creation date . . . : 2023/03/25     
   Data set encryption : NO              Referenced date . . : 2023/03/25     
                                         Expiration date . . : ***None***     
                                                                              
   SMS Compressible  . : NO                                                   
                                                                              
  

Below is the log of the playbook:                                                                          

 

Administrator@t-2799-k ~/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch
$ ansible-playbook -i inventories copy-sort-fetch.yml -v
Using /home/Administrator/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch/ansible.cfg as config file

PLAY [source_system] **************************************************************************************


TASK [Copy local file containing data] ********************************************************************


TASK [copy-local-data : Generate a temporary data set name] ***********************************************

changed: [zos_host] => {"changed": true, "cmd": ["mvstmp", "IBMUSER"], "delta": "0:00:00.331961", "end": "2023-03-25 14:49:01.180619", "rc": 0, "start": "2023-03-25 14:49:00.848658", "stderr": "", "stderr_lines": [], "stdout": "IBMUSER.P6843330.T0172681.C0000000", "stdout_lines": ["IBMUSER.P6843330.T0172681.C0000000"]}

TASK [copy-local-data : set_fact] *************************************************************************

ok: [zos_host] => {"ansible_facts": {"dest_tmp_data_set": "IBMUSER.P6843330.T0172681.C0000000"}, "changed": false}

TASK [copy-local-data : Create the temporary data set] ****************************************************

changed: [zos_host] => {"changed": true, "message": "", "names": ["IBMUSER.P6843330.T0172681.C0000000"]}

TASK [copy-local-data : Copy local data file to remote z/OS temporary data set] ***************************

ok: [zos_host] => {"changed": null, "dest": "IBMUSER.P6843330.T0172681.C0000000", "is_binary": false, "src": "/home/Administrator/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch/roles/copy-local-data/files/SalaryData.txt"}

TASK [copy-local-data : Ensure the copied data set exists on target] **************************************

ok: [zos_host] => {"changed": false, "data_sets": [{"name": "IBMUSER.P6843330.T0172681.C0000000", "type": "NONVSAM"}], "examined": 0, "matched": 1}

TASK [copy-local-data : assert] ***************************************************************************

ok: [zos_host] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Remove the column headers so that we are left with raw data] ****************************************

changed: [zos_host] => {"changed": 1, "cmd": "dsedhelper -d -c IBM-1047 /^ case.*/d IBMUSER.P6843330.T0172681.C0000000 ", "found": 1}

TASK [Pause playbook execution] ***************************************************************************

[Pause playbook execution]
Press enter to continue, Ctrl+C to interrupt:
ok: [zos_host] => {"changed": false, "delta": 1697, "echo": true, "rc": 0, "start": "2023-03-25 09:48:56.610553", "stderr": "", "stdout": "Paused for 28.29 minutes", "stop": "2023-03-25 10:17:13.741657", "user_input": ""}

TASK [Sort the data set based on salary in descending order] **********************************************

changed: [zos_host] => {"backups": [], "changed": true, "dd_names": [{"byte_count": 1168, "content": ["\f
           SORT FIELDS=(57,9,BI,D)", "", "ICE805I 1 JOBNAME: IBMUSER5 , STEPNAME: STEP1", "ICE802I 0 BLOCKSET     TECHNIQUE IN CONTROL", "ICE906I 0 ST=ABOVE,SR=6291456,RC=0", "ICE907I 0 ST=ABOVE,SA=6291440,NF=1,LF=6291440,SF=6291440", "ICE906I 0 ST=BELOW,SR=1024000,RC=0", "ICE907I 0 ST=BELOW,SA=1023984,NF=1,LF=1023984,SF=1023984", "ICE751I 0 C5-NONE   C6-NONE   C7-I76950 C8-I76518 E4-NONE   C9-NONE   E5-I77614 E7-I76950", "ICE750I 0 DC 6140 TC 0 CS DSVII KSZ 13 VSZ 13", "ICE752I 0 FSZ=6140 BC  IGN=0 E  AVG=514 0  WSP=7 C  DYN=0 0", "ICE915I 0 MOFSZ=0,MOSZ=0,MOSYS=1501(5),MOSTG=1501,MEML=1932(0)", "ICE916I 0 MOFR=0002,MOVR=II", "ICE996I 0 ESM=613549,ESO=229281,ESR=61354,ESP=4096,ESS=16384,CES=1019136,HSZ=524288", "ICE997I 0 HWSP=0,HMAX=384267,MOMAX=0,ASV=384267,EQ=N0,HN=0", "ICE898I 0 OMAX=154986,NMAX=613549,ENQT=384267,CMAX=0,HU=0,BUN=0,MD=N3,N3,DU=0,DR=0,HN=0", "ICE889I 0 CT=MAX     , SB=8, L=0, D=0000, CCW=1MAM", "ICE902I 0 O PP10  I PP11", "ICE906I 1 ST=ABOVE,SR=6234096,RC=0", "ICE907I 1 ST=ABOVE,SA=6234080,NF=1,LF=6234080,SF=6234080", "ICE906I 1 ST=BELOW,SR=57144,RC=0", "ICE907I 1 ST=BELOW,SA=48936,NF=1,LF=48936,SF=48936", "ICE218A 0 5 BYTE VARIABLE RECORD IS SHORTER THAN 65 BYTE MINIMUM FOR          FIELDS", ""], "dd_name": "sysout", "name": "IBMUSER.P0066131.T0778115.C0000000", "record_count": 24}], "ret_code": {"code": 16}}

TASK [Fetch the sorted data set back to controller] *******************************************************

changed: [zos_host] => {"changed": true, "checksum": "42d8f38dd2e0105f7795e79ebd11122dae12cf2e2b1e8fed709548e2343535db", "data_set_type": "Sequential", "dest": "/home/Administrator/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch/SortedSalaryData.txt", "file": "IBMUSER.P6843330.T0172681.C0000000", "is_binary": false}

TASK [Add column headers back to the sorted file] *********************************************************

changed: [zos_host] => {"backup": "", "changed": true, "msg": "line added"}

TASK [Pause playbook execution] ***************************************************************************

_____________________________________________________________________________________________________________

i was able to fix the issue by updating the code block dd_input  in copy-sort-fetch.yml  to add "OPTION VLSHRT" as shown below

 

 - dd_input:
                dd_name: sysin
                content:
                  - "   OPTION VLSHRT "
                  - "   SORT FIELDS=(57,9,BI,D) "

 

now the playbook runs successfully with this output:

Administrator@t-2799-k ~/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch
$ ansible-playbook -i inventories copy-sort-fetch.yml -v
Using /home/Administrator/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch/ansible.cfg as config file

PLAY [source_system] **************************************************************************************


TASK [Copy local file containing data] ********************************************************************


TASK [copy-local-data : Generate a temporary data set name] ***********************************************

changed: [zos_host] => {"changed": true, "cmd": ["mvstmp", "IBMUSER"], "delta": "0:00:00.359676", "end": "2023-03-25 17:23:51.877078", "rc": 0, "start": "2023-03-25 17:23:51.517402", "stderr": "", "stderr_lines": [], "stdout": "IBMUSER.P0066634.T0869093.C0000000", "stdout_lines": ["IBMUSER.P0066634.T0869093.C0000000"]}

TASK [copy-local-data : set_fact] *************************************************************************

ok: [zos_host] => {"ansible_facts": {"dest_tmp_data_set": "IBMUSER.P0066634.T0869093.C0000000"}, "changed": false}

TASK [copy-local-data : Create the temporary data set] ****************************************************

changed: [zos_host] => {"changed": true, "message": "", "names": ["IBMUSER.P0066634.T0869093.C0000000"]}

TASK [copy-local-data : Copy local data file to remote z/OS temporary data set] ***************************

ok: [zos_host] => {"changed": null, "dest": "IBMUSER.P0066634.T0869093.C0000000", "is_binary": false, "src": "/home/Administrator/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch/roles/copy-local-data/files/SalaryData.txt"}

TASK [copy-local-data : Ensure the copied data set exists on target] **************************************

ok: [zos_host] => {"changed": false, "data_sets": [{"name": "IBMUSER.P0066634.T0869093.C0000000", "type": "NONVSAM"}], "examined": 0, "matched": 1}

TASK [copy-local-data : assert] ***************************************************************************

ok: [zos_host] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Pause playbook execution] ***************************************************************************

[Pause playbook execution]
Press enter to continue, Ctrl+C to interrupt:
ok: [zos_host] => {"changed": false, "delta": 6, "echo": true, "rc": 0, "start": "2023-03-25 12:23:40.100909", "stderr": "", "stdout": "Paused for 0.11 minutes", "stop": "2023-03-25 12:23:46.593762", "user_input":
""}

TASK [Remove the column headers so that we are left with raw data] ****************************************

changed: [zos_host] => {"changed": 1, "cmd": "dsedhelper -d -c IBM-1047 /^ case.*/d IBMUSER.P0066634.T0869093.C0000000 ", "found": 1}

TASK [Sort the data set based on salary in descending order] **********************************************

changed: [zos_host] => {"backups": [], "changed": true, "dd_names": [{"byte_count": 65, "content": ["\f
         OPTION VLSHRT", "             SORT FIELDS=(57,9,BI,D)", ""], "dd_name": "sysout", "name": "IBMUSER.P6843865.T0051716.C0000000", "record_count": 3}], "ret_code": {"code": 0}}

TASK [Fetch the sorted data set back to controller] *******************************************************

changed: [zos_host] => {"changed": true, "checksum": "e801546eb3714140b618c337416149fc4f396ec9ebdb70e993ddb7441e8d89b1", "data_set_type": "Sequential", "dest": "/home/Administrator/ansible-demo/z_ansible_collections_samples/zos_concepts/data_transfer/copy_sort_fetch/SortedSalaryData.txt", "file": "IBMUSER.P0066634.T0869093.C0000000", "is_binary": false}

TASK [Add column headers back to the sorted file] *********************************************************

changed: [zos_host] => {"backup": "", "changed": true, "msg": "line added"}

TASK [Pause playbook execution] ***************************************************************************

[Pause playbook execution]
Press enter to continue, Ctrl+C to interrupt:
ok: [zos_host] => {"changed": false, "delta": 2, "echo": true, "rc": 0, "start": "2023-03-25 12:24:18.146623", "stderr": "", "stdout": "Paused for 0.04 minutes", "stop": "2023-03-25 12:24:20.335701", "user_input":
""}

TASK [Clean up remote data set] ***************************************************************************

changed: [zos_host] => {"changed": true, "message": "", "names": ["IBMUSER.P0066634.T0869093.C0000000"]}

PLAY RECAP ************************************************************************************************

zos_host                   : ok=13   changed=7    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

**Comment by richp405** [comment]
I'm a bit confused since I don't see a 'pause' in the referenced file.  I'll look for variations, but I believe the file the customer referenced is not the original file.

**Comment by richp405** [comment]
Here is an image of the first replication attempt (where the run succeeded, not generating the error).  I'll look to see if there is a similarly named file in a different folder.



![Image](https://user-images.githubusercontent.com/39140439/231246614-50a31098-b139-4818-8eaf-b00a014bd86e.png)

**Comment by richp405** [comment]
I have duplicated the error: this is off the original file, with verbose (-v) enabled.  I will need to research this message.


![Image](https://user-images.githubusercontent.com/39140439/231250941-1623dec8-c921-45c7-8e26-e805f221b308.png)

**Comment by richp405** [comment]
Note was sent to UK team, but they were in a new rollout, so it will be another week or so before we can change the zTrial master copy.  I'm looking to see about making the change in the demo archive.

**Comment by richp405** [comment]
I've created richp405-patch-2 in the samples repository, but can't get it to merge because of issues with ansible-lint, which in turn can't import openstack tools.  @ddimatos

**Comment by richp405** [comment]
Actual pr: https://github.com/IBM/z_ansible_collections_samples/pull/208

**Comment by richp405** [comment]
Got edit into zTrial, where Biao merged into prod.  I pulled a new copy, and the changed files are there.  Ran the playbook, and it went without error.  Closing issue.
